pgx/src/datum/range.rs

```diff
@@ -9,56 +9,281 @@ Use of this source code is governed by the MIT license that can be found in the
 
 //! Utility functions for working with `pg_sys::RangeType` structs
 use crate::{
-    pg_sys, void_mut_ptr, AnyNumeric, Date, FromDatum, IntoDatum, Numeric, Timestamp,
-    TimestampWithTimeZone,
+    pg_sys, AnyNumeric, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
 };
-use pgx_pg_sys::{Oid, RangeBound};
 use pgx_sql_entity_graph::metadata::{
     ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
 };
-use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive};
+
+/// A Postgres Range's "lower" or "upper" value
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum RangeBound<T> {
+    Infinite,
+    Inclusive(T),
+    Exclusive(T),
+}
 
-/// Represents Datum to serialized RangeType PG struct
-pub struct Range<T: FromDatum + IntoDatum + RangeSubType> {
-    ptr: *mut pg_sys::varlena,
-    range_type: *mut pg_sys::RangeType,
-    _marker: PhantomData<T>,
+impl<T> RangeBound<T>
+where
+    T: RangeSubType,
+{
+    /// Returns a reference to this [`RangeBound`]'s underlying value.
+    ///
+    /// If this [`RangeBound`] is [`RangeBound::Infinite`], then `None` is returned, otherwise
+    /// `Some(&T)`
+    #[inline]
+    pub fn get(&self) -> Option<&T> {
+        match self {
+            RangeBound::Infinite => None,
+            RangeBound::Inclusive(v) | RangeBound::Exclusive(v) => Some(v),
+        }
+    }
+
+    /// Convert this pgx [`RangeBound`] into the equivalent Postgres [`pg_sys::RangeBound`].
+    ///
+    /// Note that the `lower` property is always set to false as a [`RangeBound`] doesn't know the
+    /// end on which it's placed.
+    pub fn into_pg(self) -> pg_sys::RangeBound {
+        match self {
+            RangeBound::Infinite => pg_sys::RangeBound {
+                val: pg_sys::Datum::from(0),
+                infinite: true,
+                inclusive: false,
+                lower: false,
+            },
+            RangeBound::Inclusive(v) => pg_sys::RangeBound {
+                val: v.into_datum().unwrap(),
+                infinite: false,
+                inclusive: true,
+                lower: false,
+            },
+            RangeBound::Exclusive(v) => pg_sys::RangeBound {
+                val: v.into_datum().unwrap(),
+                infinite: false,
+                inclusive: false,
+                lower: false,
```

Comment:
Considering the repeated logic here, do you think it would make sense for pgx-pg-sys to contain an `impl Default for pg_sys::RangeBound`? `pg_sys::Datum` doesn't have an impl for Default because pointers don't, but in this case it feels correct to have the default be the same as `RangeBound::Infinite` here.

Then these could reuse that like `pg_sys::RangeBound { val: v.into_datum().unwrap(), infinite: false, ..Default::default() }`. Though that's only saving us a few lines and may require cross-reference so... maybe not.

Reply:
all the bindgen generated structs already `#[derive(Default])`, which of course will set `infinite` to `false`.  Not sure it's worth special-casing this one thing.

Reply:
...wow you're right, that sure is... an impl of `Default` there.

Reply:
...I guess we can uhhh, use... that... then...
**wow**.